Lib/tempfile.py

```diff
@@ -446,16 +446,19 @@ def close(self, unlink=_os.unlink):
                     if self.delete and self.delete_on_close:
                         unlink(self.name)
 
-        # Need to ensure the file is deleted on __del__
-        def __del__(self):
-            self.close()
 
     else:
         def close(self):
             if not self.close_called:
                 self.close_called = True
                 self.file.close()
 
+    def __del__(self):
+        self.close()
+        # Need to ensure the file is deleted on __del__ if delete = True and
+        # file still exists
+        if self.delete and _os.path.exists(self.name):
+            _os.unlink(self.name)
```

Comment:
Implementing this here is a mistake. Deleting the file in `_TemporaryFileCloser` should only be due to `delete_on_close`. It should not try to delete the file if it was already deleted. We don't know whether an unrelated file was created with the same name. In particular, `_TemporaryFileCloser.__del__` should not try to delete the file again if it's handled by `close()`, i.e. if `delete_on_close` is true, and it should not try to delete the file again if `__exit__()` was called. But how does `_TemporaryFileCloser` know whether `_TemporaryFileWrapper.__exit__()` was called? The file being closed is not enough information, since `close()` may have been called manually. It is not the job of `_TemporaryFileCloser` to clean up the file if `delete` is true and `delete_on_close` is false. That's the job of `_TemporaryFileWrapper`.

Reply:
I acknowledged I moved the code to a different place, than you suggested, but I saw, that `__del__()` is already captured in the `_TemporaryFileCloser` (see [here ](https://github.com/python/cpython/blob/80bc7d7c0a3f777da244ec79429baf382523b8da/Lib/tempfile.py#L448)

I also noticed, that `__del__()` in  `_TemporaryFileCloser` is called after the `__del__()` on `_TemporaryFileWrapper` (which was creating a problem with the tests, when run on Linux, which are gone now).

Any way, in your suggestion, we will have to do something with existing `__del__` code in the `_TemporaryFileCloser` (see [here ](https://github.com/python/cpython/blob/80bc7d7c0a3f777da244ec79429baf382523b8da/Lib/tempfile.py#L448) to avoid the problem in POSIX

Reply:
`_TemporaryFileCloser.__del__()` is only there to ensure that the file gets closed. This also guarantees to delete the file if both `delete` and `delete_on_close` are true. In this case, `unlink()` is guaranteed to only get called once because the work is gated by `close_called`. This behavior must not change. 

Similarly, if `delete` is true and `delete_on_close` is false, if the file is deleted by `__exit__()`, then `_TemporaryFileWrapper.__del__()` must not try to delete the file again. That's why I added the `_cleanup_called` class attribute in my example and removed redundant work by adding a common `_cleanup()` method.

The combined behavior of `_TemporaryFileCloser` and `_TemporaryFileWrapper` must never call `unlink()` more than once.

Reply:
@eryksun , thanks!

I understand, that the combined behavior of `_TemporaryFileCloser` and `_TemporaryFileWrapper` must never call `unlink()` more than once. 

What I still don't get is the following: in the current release we already have a code, which handles finalization event in the `_TemporaryFileCloser`.  If we now add another `__del__()` into the  `_TemporaryFileWrapper`, then during finalization first the `_TemporaryFileWrapper.__del__()` will run and later the `_TemporaryFileCloser.__del__()` and we have to make sure they don't conflict with each other plus that it is clear anyone how it all works together with each other

Isn't it is better from design perspective to place the code, which handles the finalization event into one place?

Reply:
There is no conflict, and the behavior should be simple to understand. The closer only deletes the file if both `delete` and `delete_on_close` are true. The wrapper only deletes the file if `delete` is true and `delete_on_close` is false. Both cases need to be gated by variables such as `close_called` and `_cleanup_called` in order to ensure that `unlink()` is only called once. In particular, when `delete` and `delete_on_close` are true, the finalizer of the closer shouldn't try to delete the file if `close()` was already called, and if `delete` is true and `delete_on_close` is false, the finalizer of the wrapper shouldn't try to delete the file if `__exit__()` was already called.

Reply:
OK, understood, thanks.
Another, a bit different, question: where the situation will be handled, in case `delete = True` and `delete_on_close = False` and no context manager was used and file was not even explicitly closed (not a good programming).

Like this:

```py
        def my_func(dir)->str:
            f = tempfile.NamedTemporaryFile(dir=dir, delete=True, delete_on_close=False)
            tmp_name = f.name
            f.write(b'blat')
            # Testing extreme case, where the file is not even explicitly closed
            # f.close()
            return tmp_name
```
Which part of the code will be responsible to delete  the file?

Reply:
The file would be deleted by the wrapper's `__del__()` method, which in the example I wrote would call the `_cleanup()` method. In CPython, the delete will occur immediately when `my_func()` returns, so I can't imagine what reason anyone would have to return `tmp_name`, the name of a deleted file.

Reply:
@eryksun , sorry if this is a stupid question, but can you explain me please why we need **class** attributes `name = None` and `_cleanup_called = False`, why can these not be an instance attributes, initialised in `__init__` ?

In all the documentation I read they say class attributes are used if data needs to be shared by all instances. Surely the flag `_cleanup_called` we want to be specific for every instance of the temporary file, as we can have many of them (or am I badly missing something)? 

I see the similar approach also being used in the `_TemporaryFileCloser`

Reply:
I have now implemented this as per your suggestions . There are 2 things however

- I still have some questions about usage of class attributes (see above)

- I had to add one if statement [here](https://github.com/Ev2geny/cpython/blob/5de68c0556058c762572e8afe4c2981b38987a1b/Lib/tempfile.py#L447):

```py
                    if self.delete and self.delete_on_close:
                        if _os.path.exists(self.name):
                            unlink(self.name)
```

Otherwise the tests were not passing for test_tempfile
- There are some errors now introduced in other parts of Python. So I think will need your advice what could be wrong

<details>

  <summary>Click me</summary>

== CPython 3.12.0a0 (heads/fix-issue-14243:0e0b6ba04d, Sep 23 2022, 11:54:11) [GCC 11.2.0]

== Linux-5.15.0-48-generic-x86_64-with-glibc2.35 little-endian

== cwd: /home/evg/evpy/cpython/build/test_python_6764æ

== CPU count: 3

== encodings: locale=UTF-8, FS=utf-8

0:00:00 load avg: 0.30 Run tests sequentially

0:00:00 load avg: 0.30 [1/1] test_tempfile

test_nonempty_list (test.test_tempfile.TestCandidateTempdirList.test_nonempty_list) ... ok

test_wanted_dirs (test.test_tempfile.TestCandidateTempdirList.test_wanted_dirs) ... ok

test_exports (test.test_tempfile.TestExports.test_exports) ... ok

test_retval (test.test_tempfile.TestGetCandidateNames.test_retval) ... ok

test_same_thing (test.test_tempfile.TestGetCandidateNames.test_same_thing) ... ok

test_no_files_left_behind (test.test_tempfile.TestGetDefaultTempdir.test_no_files_left_behind) ... ok

test_case_sensitive (test.test_tempfile.TestGetTempDir.test_case_sensitive) ... ok

test_directory_exists (test.test_tempfile.TestGetTempDir.test_directory_exists) ... ok

test_directory_writable (test.test_tempfile.TestGetTempDir.test_directory_writable) ... ok

test_same_thing (test.test_tempfile.TestGetTempDir.test_same_thing) ... ok

test_sane_template (test.test_tempfile.TestGetTempPrefix.test_sane_template) ... ok

test_usable_template (test.test_tempfile.TestGetTempPrefix.test_usable_template) ... ok

test_infer_return_type_multiples (test.test_tempfile.TestLowLevelInternals.test_infer_return_type_multiples) ... ok

test_infer_return_type_multiples_and_none (test.test_tempfile.TestLowLevelInternals.test_infer_return_type_multiples_and_none) ... ok

test_infer_return_type_pathlib (test.test_tempfile.TestLowLevelInternals.test_infer_return_type_pathlib) ... ok

test_infer_return_type_pathlike (test.test_tempfile.TestLowLevelInternals.test_infer_return_type_pathlike) ... ok

test_infer_return_type_singles (test.test_tempfile.TestLowLevelInternals.test_infer_return_type_singles) ... ok

test_basic (test.test_tempfile.TestMkdtemp.test_basic) ... ok

test_basic_many (test.test_tempfile.TestMkdtemp.test_basic_many) ... ok

test_basic_with_bytes_names (test.test_tempfile.TestMkdtemp.test_basic_with_bytes_names) ... ok

test_choose_directory (test.test_tempfile.TestMkdtemp.test_choose_directory) ... ok

test_collision_with_existing_directory (test.test_tempfile.TestMkdtemp.test_collision_with_existing_directory) ... ok

test_collision_with_existing_file (test.test_tempfile.TestMkdtemp.test_collision_with_existing_file) ... ok

test_for_tempdir_is_bytes_issue40701_api_warts (test.test_tempfile.TestMkdtemp.test_for_tempdir_is_bytes_issue40701_api_warts) ... ok

test_mode (test.test_tempfile.TestMkdtemp.test_mode) ... ok

test_non_directory (test.test_tempfile.TestMkdtemp.test_non_directory) ... ok

test_nonexisting_directory (test.test_tempfile.TestMkdtemp.test_nonexisting_directory) ... ok

test_read_only_directory (test.test_tempfile.TestMkdtemp.test_read_only_directory) ... ok

test_basic (test.test_tempfile.TestMkstemp.test_basic) ... ok

test_basic_with_bytes_names (test.test_tempfile.TestMkstemp.test_basic_with_bytes_names) ... ok

test_choose_directory (test.test_tempfile.TestMkstemp.test_choose_directory) ... ok

test_for_tempdir_is_bytes_issue40701_api_warts (test.test_tempfile.TestMkstemp.test_for_tempdir_is_bytes_issue40701_api_warts) ... ok

test_basic (test.test_tempfile.TestMkstempInner.test_basic) ... ok

test_basic_many (test.test_tempfile.TestMkstempInner.test_basic_many) ... ok

test_basic_with_bytes_names (test.test_tempfile.TestMkstempInner.test_basic_with_bytes_names) ... ok

test_choose_directory (test.test_tempfile.TestMkstempInner.test_choose_directory) ... ok

test_collision_with_existing_directory (test.test_tempfile.TestMkstempInner.test_collision_with_existing_directory) ... ok

test_collision_with_existing_file (test.test_tempfile.TestMkstempInner.test_collision_with_existing_file) ... ok

test_file_mode (test.test_tempfile.TestMkstempInner.test_file_mode) ... ok

test_noinherit (test.test_tempfile.TestMkstempInner.test_noinherit) ... ok

test_non_directory (test.test_tempfile.TestMkstempInner.test_non_directory) ... ok

test_nonexisting_directory (test.test_tempfile.TestMkstempInner.test_nonexisting_directory) ... ok

test_read_only_directory (test.test_tempfile.TestMkstempInner.test_read_only_directory) ... ok

test_textmode (test.test_tempfile.TestMkstempInner.test_textmode) ... skipped 'text mode not available'

test_basic (test.test_tempfile.TestMktemp.test_basic) ... ok

test_many (test.test_tempfile.TestMktemp.test_many) ... ok

test_bad_encoding (test.test_tempfile.TestNamedTemporaryFile.test_bad_encoding) ... ok

test_bad_mode (test.test_tempfile.TestNamedTemporaryFile.test_bad_mode) ... ok

test_basic (test.test_tempfile.TestNamedTemporaryFile.test_basic) ... ok

test_context_man_not_del_if_delete_false (test.test_tempfile.TestNamedTemporaryFile.test_context_man_not_del_if_delete_false) ... ok

test_context_man_not_del_on_close_if_delete_on_close_false (test.test_tempfile.TestNamedTemporaryFile.test_context_man_not_del_on_close_if_delete_on_close_false) ... ok

test_context_man_ok_to_delete_manually (test.test_tempfile.TestNamedTemporaryFile.test_context_man_ok_to_delete_manually) ... ok

test_context_manager (test.test_tempfile.TestNamedTemporaryFile.test_context_manager) ... ok

test_correct_finalizer_work_if_already_deleted (test.test_tempfile.TestNamedTemporaryFile.test_correct_finalizer_work_if_already_deleted) ... ok

test_creates_named (test.test_tempfile.TestNamedTemporaryFile.test_creates_named) ... ok

test_del_by_finalizer (test.test_tempfile.TestNamedTemporaryFile.test_del_by_finalizer) ... ok

test_del_on_close (test.test_tempfile.TestNamedTemporaryFile.test_del_on_close) ... ok

test_dis_del_on_close (test.test_tempfile.TestNamedTemporaryFile.test_dis_del_on_close) ... ok

test_iter (test.test_tempfile.TestNamedTemporaryFile.test_iter) ... ok

test_method_lookup (test.test_tempfile.TestNamedTemporaryFile.test_method_lookup) ... ok

test_multiple_close (test.test_tempfile.TestNamedTemporaryFile.test_multiple_close) ... ok

test_unexpected_error (test.test_tempfile.TestNamedTemporaryFile.test_unexpected_error) ... ok

test_get_eight_char_str (test.test_tempfile.TestRandomNameSequence.test_get_eight_char_str) ... ok

test_many (test.test_tempfile.TestRandomNameSequence.test_many) ... ok

test_process_awareness (test.test_tempfile.TestRandomNameSequence.test_process_awareness) ... ok

test_basic (test.test_tempfile.TestSpooledTemporaryFile.test_basic) ... ok

test_bound_methods (test.test_tempfile.TestSpooledTemporaryFile.test_bound_methods) ... ok

test_class_getitem (test.test_tempfile.TestSpooledTemporaryFile.test_class_getitem) ... ok

test_context_manager_after_rollover (test.test_tempfile.TestSpooledTemporaryFile.test_context_manager_after_rollover) ... ok

test_context_manager_before_rollover (test.test_tempfile.TestSpooledTemporaryFile.test_context_manager_before_rollover) ... ok

test_context_manager_during_rollover (test.test_tempfile.TestSpooledTemporaryFile.test_context_manager_during_rollover) ... ok

test_del_on_close (test.test_tempfile.TestSpooledTemporaryFile.test_del_on_close) ... ok

test_del_rolled_file (test.test_tempfile.TestSpooledTemporaryFile.test_del_rolled_file) ... ok

test_del_unrolled_file (test.test_tempfile.TestSpooledTemporaryFile.test_del_unrolled_file) ... ok

test_fileno (test.test_tempfile.TestSpooledTemporaryFile.test_fileno) ... ok

test_iobase_interface (test.test_tempfile.TestSpooledTemporaryFile.test_iobase_interface) ... ok

test_is_iobase (test.test_tempfile.TestSpooledTemporaryFile.test_is_iobase) ... ok

test_multiple_close_after_rollover (test.test_tempfile.TestSpooledTemporaryFile.test_multiple_close_after_rollover) ... ok

test_multiple_close_before_rollover (test.test_tempfile.TestSpooledTemporaryFile.test_multiple_close_before_rollover) ... ok

test_properties (test.test_tempfile.TestSpooledTemporaryFile.test_properties) ... ok

test_rewrite_small (test.test_tempfile.TestSpooledTemporaryFile.test_rewrite_small) ... ok

test_sparse (test.test_tempfile.TestSpooledTemporaryFile.test_sparse) ... ok

test_text_mode (test.test_tempfile.TestSpooledTemporaryFile.test_text_mode) ... ok

test_text_newline_and_encoding (test.test_tempfile.TestSpooledTemporaryFile.test_text_newline_and_encoding) ... ok

test_truncate_with_size_parameter (test.test_tempfile.TestSpooledTemporaryFile.test_truncate_with_size_parameter) ... ok

test_write_sequential (test.test_tempfile.TestSpooledTemporaryFile.test_write_sequential) ... ok

test_writelines (test.test_tempfile.TestSpooledTemporaryFile.test_writelines) ... ok

test_writelines_sequential (test.test_tempfile.TestSpooledTemporaryFile.test_writelines_sequential) ... ok

test_cleanup_with_symlink_to_a_directory (test.test_tempfile.TestTemporaryDirectory.test_cleanup_with_symlink_to_a_directory) ... ok

test_context_manager (test.test_tempfile.TestTemporaryDirectory.test_context_manager) ... ok

test_del_on_collection (test.test_tempfile.TestTemporaryDirectory.test_del_on_collection) ... ok

test_del_on_collection_ignore_errors (test.test_tempfile.TestTemporaryDirectory.test_del_on_collection_ignore_errors)

Test that ignoring errors works when TemporaryDirectory is gced. ... ok

test_del_on_shutdown (test.test_tempfile.TestTemporaryDirectory.test_del_on_shutdown) ... ok

test_del_on_shutdown_ignore_errors (test.test_tempfile.TestTemporaryDirectory.test_del_on_shutdown_ignore_errors)

Test ignoring errors works when a tempdir is gc'ed on shutdown. ... ok

test_exit_on_shutdown (test.test_tempfile.TestTemporaryDirectory.test_exit_on_shutdown) ... ok

test_explicit_cleanup (test.test_tempfile.TestTemporaryDirectory.test_explicit_cleanup) ... ok

test_explict_cleanup_ignore_errors (test.test_tempfile.TestTemporaryDirectory.test_explict_cleanup_ignore_errors)

Test that cleanup doesn't return an error when ignoring them. ... ok

test_flags (test.test_tempfile.TestTemporaryDirectory.test_flags) ... skipped 'requires os.lchflags'

test_mkdtemp_failure (test.test_tempfile.TestTemporaryDirectory.test_mkdtemp_failure) ... ok

test_modes (test.test_tempfile.TestTemporaryDirectory.test_modes) ... ok

test_multiple_close (test.test_tempfile.TestTemporaryDirectory.test_multiple_close) ... ok

test_warnings_on_cleanup (test.test_tempfile.TestTemporaryDirectory.test_warnings_on_cleanup) ... ok

test_bad_encoding (test.test_tempfile.TestTemporaryFile.test_bad_encoding) ... ok

test_bad_mode (test.test_tempfile.TestTemporaryFile.test_bad_mode) ... ok

test_basic (test.test_tempfile.TestTemporaryFile.test_basic) ... ok

test_has_no_name (test.test_tempfile.TestTemporaryFile.test_has_no_name) ... ok

test_mode_and_encoding (test.test_tempfile.TestTemporaryFile.test_mode_and_encoding) ... ok

test_multiple_close (test.test_tempfile.TestTemporaryFile.test_multiple_close) ... ok

test_unexpected_error (test.test_tempfile.TestTemporaryFile.test_unexpected_error) ... ok



----------------------------------------------------------------------

Ran 109 tests in 1.041s



OK (skipped=2)

test_tempfile failed (env changed)



== Tests result: SUCCESS ==



1 test altered the execution environment:

    test_tempfile



Total duration: 1.2 sec

Tests result: SUCCESS

</details>